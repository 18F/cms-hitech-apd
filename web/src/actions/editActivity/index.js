export * from './contractorResources';
export * from './costAllocate';
export * from './overview';
