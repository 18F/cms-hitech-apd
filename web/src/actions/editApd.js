import { updateBudget } from './apd';

export const ADD_APD_ITEM = Symbol('add apd item');
export const EDIT_APD = Symbol('edit apd');
export const REMOVE_APD_ITEM = Symbol('remove apd item');

export const addKeyPerson = () => (dispatch, getState) => {
  dispatch({
    type: ADD_APD_ITEM,
    path: '/keyPersonnel/-',
    state: getState()
  });
};

export const removeKeyPerson = (
  index,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this key person?')) {
    dispatch({
      type: REMOVE_APD_ITEM,
      path: `/keyPersonnel/${index}`
    });
    dispatch(updateBudget());
  }
};

export const setKeyPersonName = (name, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/name`,
  value: name
});

export const setKeyPersonEmail = (email, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/email`,
  value: email
});

export const setKeyPersonRole = (role, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/position`,
  value: role
});

export const setKeyPersonPercentTime = (percent, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/percentTime`,
  value: percent
});

// needs to trigger budget calculation
export const setKeyPersonHasCosts = (hasCosts, index) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/hasCosts`,
    value: hasCosts
  });
  dispatch(updateBudget());
};

// needs to trigger budget calculation
export const setKeyPersonCost = (cost, index, year) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/costs/${year}`,
    value: cost
  });
  dispatch(updateBudget());
};

export const setMedicaidDirectorEmail = email => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/email',
    value: email
  };
};

export const setMedicaidDirectorName = name => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/name',
    value: name
  };
};

export const setMedicaidDirectorPhoneNumber = phone => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/phone',
    value: phone
  };
};

export const setMedicaidOfficeAddress1 = address => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/address1',
    value: address
  };
};

export const setMedicaidOfficeAddress2 = address => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/address2',
    value: address
  };
};

export const setMedicaidOfficeCity = city => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/city',
    value: city
  };
};

export const setMedicaidOfficeState = state => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/state',
    value: state
  };
};

export const setMedicaidOfficeZip = zip => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/zip',
    value: zip
  };
};
