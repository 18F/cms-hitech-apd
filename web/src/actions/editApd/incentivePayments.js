import { EDIT_APD } from './symbols';

export const setIncentiveEHCount = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/ehCt/${year}/${quarter}`,
  value
});

export const setIncentiveEHPayment = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/ehAmt/${year}/${quarter}`,
  value
});

export const setIncentiveEPCount = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/epCt/${year}/${quarter}`,
  value
});

export const setIncentiveEPPayment = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/epAmt/${year}/${quarter}`,
  value
});
