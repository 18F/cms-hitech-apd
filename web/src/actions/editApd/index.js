export * from './symbols';

export * from './keyPersonnel';
export * from './overview';
export * from './previousActivities';
export * from './stateProfile';
