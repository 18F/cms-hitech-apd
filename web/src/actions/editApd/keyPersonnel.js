import { ADD_APD_ITEM, EDIT_APD, REMOVE_APD_ITEM } from './symbols';
import { updateBudget } from '../apd';

export const addKeyPerson = () => (dispatch, getState) => {
  dispatch({
    type: ADD_APD_ITEM,
    path: '/keyPersonnel/-',
    state: getState()
  });
};

export const removeKeyPerson = (
  index,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this key person?')) {
    dispatch({
      type: REMOVE_APD_ITEM,
      path: `/keyPersonnel/${index}`
    });
    dispatch(updateBudget());
  }
};

export const setKeyPersonName = (name, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/name`,
  value: name
});

export const setKeyPersonEmail = (email, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/email`,
  value: email
});

export const setKeyPersonRole = (role, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/position`,
  value: role
});

export const setKeyPersonPercentTime = (percent, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/percentTime`,
  value: percent
});

// needs to trigger budget calculation
export const setKeyPersonHasCosts = (hasCosts, index) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/hasCosts`,
    value: hasCosts
  });
  dispatch(updateBudget());
};

// needs to trigger budget calculation
export const setKeyPersonCost = (cost, index, year) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/costs/${year}`,
    value: cost
  });
  dispatch(updateBudget());
};
