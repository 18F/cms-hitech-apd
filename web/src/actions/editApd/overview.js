import { ADD_APD_YEAR, EDIT_APD, REMOVE_APD_YEAR } from './symbols';
import { updateBudget } from '../apd';

export const addYear = year => (dispatch, getState) => {
  dispatch({ type: ADD_APD_YEAR, value: year, state: getState() });
  dispatch(updateBudget());
};

export const removeYear = year => (dispatch, getState) => {
  dispatch({ type: REMOVE_APD_YEAR, value: year, state: getState() });
  dispatch(updateBudget());
};

export const setNarrativeForHIE = text => ({
  type: EDIT_APD,
  path: '/narrativeHIE',
  value: text
});

export const setNarrativeForHIT = text => ({
  type: EDIT_APD,
  path: '/narrativeHIT',
  value: text
});

export const setNarrativeForMMIS = text => ({
  type: EDIT_APD,
  path: '/narrativeMMIS',
  value: text
});

export const setProgramOverview = text => ({
  type: EDIT_APD,
  path: '/programOverview',
  value: text
});
