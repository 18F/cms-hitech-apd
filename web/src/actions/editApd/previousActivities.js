import { EDIT_APD } from './symbols';

export const setPreviousActivitySummary = summary => ({
  type: EDIT_APD,
  path: '/previousActivitySummary',
  value: summary
});

export const setPreviousActivityApprovedExpenseForHITandHIE = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/hithie/totalApproved`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseForHITandHIE = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/hithie/federalActual`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS50FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/50/totalApproved`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS75FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/75/totalApproved`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS90FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/90/totalApproved`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS50FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/50/federalActual`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS75FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/75/federalActual`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS90FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/90/federalActual`,
  value: expense
});
