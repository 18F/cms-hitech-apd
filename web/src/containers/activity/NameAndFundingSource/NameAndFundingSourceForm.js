import { ChoiceList, TextField } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment, useCallback } from 'react';
import { useDispatch } from 'react-redux';

import {
  setActivityName,
  setActivityFundingSource
} from '../../../actions/editActivity';

const NameAndFundingSourceForm = ({ index, item: { fundingSource, name } }) => {
  const dispatch = useDispatch();

  const changeName = useCallback(({ target: { value } }) => {
    // These forms are built from the list of editable activities, which is
    // all activities except the first one. So the local list index starts at
    // 0, but it corresponds to index 1 in the full list. To handle that, just
    // increment the local index by one when updating the activity.
    dispatch(setActivityName(index + 1, value));
  });

  const changeFundingSource = useCallback(({ target: { value } }) => {
    // Same thing here.
    dispatch(setActivityFundingSource(index + 1, value));
  });

  return (
    <Fragment>
      <TextField
        autoFocus="true"
        label="Activity name"
        name="activity-name"
        value={name}
        onChange={changeName}
      />
      <ChoiceList
        label="Program type"
        name="program-type"
        choices={[
          { checked: fundingSource === 'HIT', label: 'HIT', value: 'HIT' },
          { checked: fundingSource === 'HIE', label: 'HIE', value: 'HIE' },
          { checked: fundingSource === 'MMIS', label: 'MMIS', value: 'MMIS' }
        ]}
        onChange={changeFundingSource}
      />
    </Fragment>
  );
};

NameAndFundingSourceForm.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    fundingSource: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired
};

export default NameAndFundingSourceForm;
