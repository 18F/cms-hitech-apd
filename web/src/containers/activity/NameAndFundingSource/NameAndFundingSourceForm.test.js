import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { UPDATE_BUDGET } from '../../../actions/apd';
import { EDIT_APD } from '../../../actions/editApd';

import NameAndFundingSource from './NameAndFundingSourceForm';

const mockStore = configureStore([thunk]);

describe('the activity name and funding source component', () => {
  const props = {
    index: 1,
    item: {
      fundingSource: 'Uncle Scrooge',
      key: 'key 1',
      name: 'Buying bikes for Huey, Dewey, and Louie'
    }
  };

  const store = mockStore('test state');

  beforeEach(() => {
    store.clearActions();
  });

  it('renders correctly', () => {
    const component = mount(
      <Provider store={store}>
        <NameAndFundingSource {...props} />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('handles changing the activity name', () => {
    const component = mount(
      <Provider store={store}>
        <NameAndFundingSource {...props} />
      </Provider>
    );

    component.find('TextField').prop('onChange')({
      target: { value: 'new value' }
    });

    expect(store.getActions()).toEqual([
      // The index that we pass via props is 1, but the name-and-funding-source
      // forms are built from a list of only the editable activities. Index 0
      // in the editable list is index 1 in the list of all activities. To
      // account for that, the form will increment the index by 1. Thus, when
      // 1 gets passed in, we expect that to become 2 when an edit happens.
      { type: EDIT_APD, path: '/activities/2/name', value: 'new value' }
    ]);
  });

  it('handles changing the funding source/program type', () => {
    const component = mount(
      <Provider store={store}>
        <NameAndFundingSource {...props} />
      </Provider>
    );

    component.find('ChoiceList').prop('onChange')({
      target: { value: 'new value' }
    });

    expect(store.getActions()).toEqual([
      {
        type: EDIT_APD,
        path: '/activities/2/fundingSource',
        value: 'new value'
      },
      { type: UPDATE_BUDGET, state: 'test state' }
    ]);
  });
});
