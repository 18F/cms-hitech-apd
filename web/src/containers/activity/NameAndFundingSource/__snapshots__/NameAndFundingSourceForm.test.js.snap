// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the activity name and funding source component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NameAndFundingSourceForm
    index={1}
    item={
      Object {
        "fundingSource": "Uncle Scrooge",
        "key": "key 1",
        "name": "Buying bikes for Huey, Dewey, and Louie",
      }
    }
  >
    <TextField
      autoFocus="true"
      label="Activity name"
      name="activity-name"
      onChange={[Function]}
      type="text"
      value="Buying bikes for Huey, Dewey, and Louie"
    >
      <div
        className="ds-u-clearfix"
      >
        <FormLabel
          component="label"
          fieldId="textfield_1"
        >
          <label
            className="ds-c-label"
            htmlFor="textfield_1"
          >
            <span
              className=""
            >
              Activity name
            </span>
          </label>
        </FormLabel>
        <input
          autoFocus="true"
          className="ds-c-field"
          id="textfield_1"
          name="activity-name"
          onChange={[Function]}
          type="text"
          value="Buying bikes for Huey, Dewey, and Louie"
        />
      </div>
    </TextField>
    <ChoiceList
      choices={
        Array [
          Object {
            "checked": false,
            "label": "HIT",
            "value": "HIT",
          },
          Object {
            "checked": false,
            "label": "HIE",
            "value": "HIE",
          },
          Object {
            "checked": false,
            "label": "MMIS",
            "value": "MMIS",
          },
        ]
      }
      label="Program type"
      name="program-type"
      onChange={[Function]}
    >
      <fieldset
        className="ds-c-fieldset"
      >
        <FormLabel
          component="legend"
        >
          <legend
            className="ds-c-label"
          >
            <span
              className=""
            >
              Program type
            </span>
          </legend>
        </FormLabel>
        <Choice
          checked={false}
          inputPlacement="left"
          key="HIT"
          name="program-type"
          onChange={[Function]}
          type="radio"
          value="HIT"
        >
          <div
            aria-atomic={null}
            aria-live={null}
            aria-relevant={null}
          >
            <input
              checked={false}
              className="ds-c-choice"
              id="radio_program-type_2"
              name="program-type"
              onChange={[Function]}
              type="radio"
              value="HIT"
            />
            <FormLabel
              component="label"
              fieldId="radio_program-type_2"
            >
              <label
                className="ds-c-label"
                htmlFor="radio_program-type_2"
              >
                <span
                  className=""
                >
                  HIT
                </span>
              </label>
            </FormLabel>
          </div>
        </Choice>
        <Choice
          checked={false}
          inputPlacement="left"
          key="HIE"
          name="program-type"
          onChange={[Function]}
          type="radio"
          value="HIE"
        >
          <div
            aria-atomic={null}
            aria-live={null}
            aria-relevant={null}
          >
            <input
              checked={false}
              className="ds-c-choice"
              id="radio_program-type_3"
              name="program-type"
              onChange={[Function]}
              type="radio"
              value="HIE"
            />
            <FormLabel
              component="label"
              fieldId="radio_program-type_3"
            >
              <label
                className="ds-c-label"
                htmlFor="radio_program-type_3"
              >
                <span
                  className=""
                >
                  HIE
                </span>
              </label>
            </FormLabel>
          </div>
        </Choice>
        <Choice
          checked={false}
          inputPlacement="left"
          key="MMIS"
          name="program-type"
          onChange={[Function]}
          type="radio"
          value="MMIS"
        >
          <div
            aria-atomic={null}
            aria-live={null}
            aria-relevant={null}
          >
            <input
              checked={false}
              className="ds-c-choice"
              id="radio_program-type_4"
              name="program-type"
              onChange={[Function]}
              type="radio"
              value="MMIS"
            />
            <FormLabel
              component="label"
              fieldId="radio_program-type_4"
            >
              <label
                className="ds-c-label"
                htmlFor="radio_program-type_4"
              >
                <span
                  className=""
                >
                  MMIS
                </span>
              </label>
            </FormLabel>
          </div>
        </Choice>
      </fieldset>
    </ChoiceList>
  </NameAndFundingSourceForm>
</Provider>
`;
